Introduce the SPIFFE Trust Domain and Bundle specification

evan2645 opened the pull request:
This commit introduces the SPIFFE Trust Domain and Bundle specification.

The specification describes the semantics of SPIFFE trust domains, how
their authorities are represented, and the mechanism by which they may
be coupled together (i.e. "federation"). It has been developed by
SPIFFE's SIG-Spec community with the supervision of the SPIFFE
maintainers, the members of which are documented in `CODEOWNERS.md`.

This change includes amendments to the X509-SVID and JWT-SVID
specifications. The mechanics of those specifications remain unchanged,
however amendments were necessary to define the manner in which their
authorities are published to and consumed from a SPIFFE bundle.

Special thank you to @justinburke, @spikecurtis, and @mattmoyer for
their efforts in shaping this document.

Signed-off-by: Evan Gilman <evan@scytale.io>

evan2645 committed:
Introduce the SPIFFE Trust Domain and Bundle specification
This commit introduces the SPIFFE Trust Domain and Bundle specification.

The specification describes the semantics of SPIFFE trust domains, how
their authorities are represented, and the mechanism by which they may
be coupled together (i.e. "federation"). It has been developed by
SPIFFE's SIG-Spec community with the supervision of the SPIFFE
maintainers, the members of which are documented in `CODEOWNERS.md`.

This change includes amendments to the X509-SVID and JWT-SVID
specifications. The mechanics of those specifications remain unchanged,
however amendments were necessary to define the manner in which their
authorities are published to and consumed from a SPIFFE bundle.

Special thank you to @justinburke, @spikecurtis, and @mattmoyer for
their efforts in shaping this document.

Signed-off-by: Evan Gilman <evan@scytale.io>

evan2645 commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+# The SPIFFE Trust Domain and Bundle
```
In the course of porting this document to markdown, I realized that none of the SPIFFE specs have authors listed. I will submit another PR once this one is in, retroactively crediting authors involved in each of the SPIFFE documents

evan2645 commented on standards/SPIFFE-ID.md:
```diff
@@ -15,7 +15,7 @@ For more general information about SPIFFE, please see the [Secure Production Ide
 2\. [SPIFFE Identity](#2-spiffe-identity)  
 2.1. [Trust Domain](#21-trust-domain)  
 2.2. [Path](#22-path)  
-2.3. [Maximum SPIFFE ID Length](#23-maximum-spiffe-id-length)
```
This is a small formatting fix to re-align the ToC following a recent change

justinburke commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+### 6.3. Using Web PKI on a SPIFFE Bundle Endpoint
+Web PKI provides a convenient way to authenticate remote SPIFFE bundle endpoints as publicly trusted authorities can negate the need for an initial out-of-band bundle exchange (as is the case when using SPIFFE authentication). While largely reliable, there are some tradeoffs and drawbacks that should be considered before its use.
+
+The first and most obvious is that you are assuming trust in the public certificate authorities and in their fulfillment procedures. Any public CA can sign a certificate for the domain name where your bundle endpoint is hosted that will appear valid. While this may be an unacceptable risk to some, most users are OK with it given the convenience it provides and the historical stability of the publ ...
```
Consider changing the second sentence to improve readability: Publicly trusted CAs have the technical ability to sign certificates which will be broadly trusted and therefore could, in theory, sign a valid certificate for the host name used to serve your bundle endpoint.

Nit: third sentence: consider s/OK/okay/.

justinburke commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+Trust bundle #1 for example.com:
+{
+	"spiffe_sequence": 1,
+        "spiffe_refresh_hint": 2419200,
```
There is an extra leading space on this line.

justinburke commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+Trust bundle #2 for example.com:
+{
+	"spiffe_sequence": 2,
+        "spiffe_refresh_hint": 2419200,
```
Extra leading space on this line, too.

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+        "spiffe_refresh_hint": 2419200,
+	"keys": [
+		{
+			"kty": "RSA"
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+		{
+			"kty": "RSA"
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #1>"]
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+			"kty": "RSA"
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #1>"]
+			"n": "<base64urlUint-encoded value>"
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+        "spiffe_refresh_hint": 2419200,
+	"keys": [
+		{
+			"kty": “RSA”
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+		{
+			"kty": “RSA”
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #1>"]
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+			"kty": “RSA”
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #1>"]
+			"n": "<base64urlUint-encoded value>"
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+			"e": "AQAB"
+		},
+		{
+			"kty": “RSA”
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+		{
+			"kty": “RSA”
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #2>"]
```
missing trailing comma

azdagron commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+			"kty": “RSA”
+			"use": "x509-svid",
+			"x5c": ["<base64 DER encoding of Certificate #2>"]
+			"n": "<base64urlUint-encoded value>"
```
missing trailing comma

spikecurtis commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+Appendix A. [SPIFFE Bundle Example](#appendix-a-spiffe-bundle-example)  
+Appendix B. [Bundle Endpoint Authentication Examples](#appendix-b-bundle-endpoint-authentication-examples)  
+Appendix B.1. [Web PKI-based Endpoint Authentication](#appendix-b1-web-pki-based-endpoint-authentication)
+Appendix B.2. [SPIFFE-based Endpoint Authentication](#appendix-b2-spiffe-based-endpoint-authentication)
```
This line is appearing on the same line as previous in the rich preview.

spikecurtis commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+
+Trust domains have a 1:N relationship with the keys that back them. A single trust domain may be represented by multiple keys and key types. For example, the former may be leveraged during root rotation, while the latter is necessary in avoiding multi-protocol attacks should more than one SVID type be in use.
+
+It should be noted that while it is possible to share cryptographic keys amongst many trust domains, it is strongly encouraged to maintain a one-to-one mapping between a trust domain and any given authoritative key. Key reuse can degrade trust domain isolation (e.g. between staging and production) and introduces additional security challenges (e.g. requiring a name constraint system for secondary ...
```
```suggestion
It should be noted that while it is possible to share cryptographic keys amongst many trust domains, we strongly advise that each authoritative key be used in a single trust domain. Key reuse can degrade trust domain isolation (e.g. between staging and production) and introduces additional security challenges (e.g. requiring a name constraint system for secondary issuers). Please see the [Security Considerations](#6-security-considerations) section for more information on this topic.
```

- passive voice sounded strange
- it's not 1:1; we just said in the previous paragraph that it's 1:N

spikecurtis commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+Note that the contents of a trust domain's bundle are expected to change over time as the keys it contains are rotated. It is the responsibility of the SPIFFE implementation to distribute bundle content updates to workloads as needed. The exact format and method by which these updates are delivered is out of scope for this specification. Please see the [SPIFFE Bundle Endpoint](#5-spiffe-bundle-en ...
+
+## 4. SPIFFE Bundle Format
+SPIFFE bundles are represented as an [RFC 7517][4] compliant JWK set. JWK was chosen for two major reasons. First, it provides a flexible format for representing various types of cryptographic keys (and documents like X.509), affording some degree of future proofing in the event that new SVID formats are defined. Second, it is widely supported and deployed, used primarily for inter-domain federat ...
```
```suggestion
SPIFFE bundles are represented as an [RFC 7517][4] compliant JWK Set. JWK was chosen for two major reasons. First, it provides a flexible format for representing various types of cryptographic keys (and documents like X.509), affording some degree of future proofing in the event that new SVID formats are defined. Second, it is widely supported and deployed, used primarily for inter-domain federation, which is a core goal of the SPIFFE project.
```

spikecurtis commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+### 6.4. Reusing Cryptographic Keys Across Trust Domains
+This specification discourages sharing cryptographic keys across trust domains since the practice degrades trust domain isolation and introduces additional security challenges. When a root key is shared across multiple trust domains, it becomes critically important that authentication and authorization implementations carefully check the trust domain component of an identity and that the trust do ...
+
+Suppose that a naïve implementation imported (ie. fully trusted) a particular root key and that the authentication system were configured to authenticate the SPIFFE identity of any SVID which chained up to the trusted root key. If the naïve implementation is not configured to only trust a specific trust domain, then any identity issued in any trust domain could be authenticated (so long as the SV ...
```
> however, this risk could alternatively be mitigated by simply not reusing root keys across trust domains.

I don't believe that not-reusing root keys across trust domains actually mitigates the example you gave (unless I'm misunderstanding the example).

Let's say my trust domain is `spike.name` and I'm federating with your trust domain, `evan.name`. Further, you and I do not share any root keys.  If I import your bundle, but my authentication system is configured to authenticate any identity that chains up to a trusted root key, it means that you can issue SVIDs for `spiffe://spike.name/admin`, and my system will accept them.  No amount of checking at the authorization layer will save me in that case.

justinburke replied:
What I'm attempting to highlight here are some of the perils in using a root key/cert for multiple trust domains. I'm trying to point out that it's insufficient to simply checking whether a certificate chains up to a trusted root; that the implementation must have an explicit trust domain to trust bundle mapping as part of the authentication process. (And I believe your example also highlights this peril.) I think the difference between what you're saying and what I wrote is that I'm not making an assumption that the operator has taken an additional step to federate with another mesh. In my example, the operator has not explicitly federated with another mesh, but has simply imported the root cert to which the certificates in the mesh chain up to. This situation could arise, for example, in an environment where the mesh does not use a CA dedicate to a single trust domain, but relies on some CA which mints certificates for multiple trust domains.

spikecurtis replied:
Yeah, that makes sense. It seems the high level point is that authentication systems really, really must check trust domain components when validating SPIFFE IDs. Hopefully we already make that point in the SVID specs, and it's useful to make again here.  There are several failure modes of failing to do this check including your example of a single CA issuing for multiple domains (without malice) and my example of a foreign CA forging SVIDs for the local trust domain.

spikecurtis reviewed (changes requested):
Couple small nits and one that might need more discussion

evan2645 committed:
Update standards/SPIFFE_Trust_Domain_and_Bundle.md

passive voice sounded strange
it's not 1:1; we just said in the previous paragraph that it's 1:N

Co-Authored-By: Spike Curtis <spike@spikecurtis.com>
Signed-off-by: Evan Gilman <evan@scytale.io>

evan2645 committed:
Update standards/SPIFFE_Trust_Domain_and_Bundle.md

Proper capitalization for JWK Set

Co-Authored-By: Spike Curtis <spike@spikecurtis.com>
Signed-off-by: Evan Gilman <evan@scytale.io>

evan2645 committed:
Minor rewording from review, JSON and space/tab formatting fixes

Signed-off-by: Evan Gilman <evan@scytale.io>

evan2645 commented:
I've accepted the suggestions and pushed a commit to address all oustanding comments with the exception of https://github.com/spiffe/spiffe/pull/106#discussion_r307938708

@spikecurtis would you mind making a suggested edit that addresses your comment there? I recall you mentioned on the last call that it is limited to the last sentence?

spikecurtis commented on standards/SPIFFE_Trust_Domain_and_Bundle.md:
```diff
@@ -0,0 +1,290 @@
+### 6.4. Reusing Cryptographic Keys Across Trust Domains
+This specification discourages sharing cryptographic keys across trust domains since the practice degrades trust domain isolation and introduces additional security challenges. When a root key is shared across multiple trust domains, it becomes critically important that authentication and authorization implementations carefully check the trust domain component of an identity and that the trust do ...
+
+Suppose that a naïve implementation imported (ie. fully trusted) a particular root key and that the authentication system were configured to authenticate the SPIFFE identity of any SVID which chained up to the trusted root key. If the naïve implementation is not configured to only trust a specific trust domain, then any identity issued in any trust domain could be authenticated (so long as the SV ...
```
```suggestion
Suppose that a naïve implementation imported (ie. fully trusted) a particular root key and that the authentication system were configured to authenticate the SPIFFE identity of any SVID which chained up to the trusted root key. If the naïve implementation is not configured to only trust a specific trust domain, then any identity issued in any trust domain could be authenticated (so long as the SVID chains up to the trusted root key).
```

I think we should just strike the last sentence, and avoid planting the seed of the idea that having each trust domain have a unique key is an appropriate alternative to robustly implemented authentication.

The message is
 a) we recommend you don't share keys across trust domains
 b) _if you are_ going to share keys, you better make damned sure that authn and authz are robust and trust-domain aware (which you should be doing anyway)

evan2645 replied:
Thanks Spike. I think that @justinburke has already stated that he's in agreement with this change so I went ahead and applied it (Justin, please chime in if that's not the case)

evan2645 committed:
Update standards/SPIFFE_Trust_Domain_and_Bundle.md

I think we should just strike the last sentence, and avoid planting the seed of the idea that having each trust domain have a unique key is an appropriate alternative to robustly implemented authentication.

Co-Authored-By: Spike Curtis <spike@spikecurtis.com>
Signed-off-by: Evan Gilman <evan@scytale.io>
